bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v3/PageImpl.java

```diff
@@ -0,0 +1,446 @@
+            return;
+        }
+
+        // Eventually add a child page that is not part page root children of the but is the entry point of the request
+        com.day.cq.wcm.api.Page entryPointPage = PageImpl.requestGetHierarchyEntryPoint(slingRequest);
+
+        if (entryPointPage == null) {
+            return;
+        }
+
+        // Filter the root page
+        if (entryPointPage.getPath().equals(currentPage.getPath())) {
+            return;
+        }
+
+        // Filter duplicates
+        if (childPages.contains(entryPointPage)) {
+            return;
+        }
+
+        childPages.add(entryPointPage);
+    }
+
+    @Nonnull
+    private Map<String, Page> getChildPageModels(@Nonnull SlingHttpServletRequest slingRequest) {
+
+        int pageTreeTraversalDepth = getPageTreeTraversalDepth(currentStyle);
+
+        List<Pattern> pageFilterPatterns = PageImpl.getStructurePatterns(slingRequest, currentStyle);
+
+        // Setting the child page to true to prevent child pages to expose their own child pages
+        SlingHttpServletRequest slingRequestWrapper = new SlingHttpServletRequestWrapper(slingRequest);
+
+        Map<String, Page> itemWrappers = new LinkedHashMap<>();
+
+        List<com.day.cq.wcm.api.Page> children = getChildPageRecursive(currentPage, slingRequestWrapper, pageFilterPatterns, pageTreeTraversalDepth);
+
+        addAsynchronousChildPage(slingRequest, children);
+
+        // Add a flag to inform the model of the child pages that they are not the root of the tree
+        slingRequestWrapper.setAttribute(ATTR_IS_CHILD_PAGE, true);
+
+        for (com.day.cq.wcm.api.Page childPage: children) {
+            Resource childPageContentResource = childPage.getContentResource();
+
+            if (childPageContentResource == null) {
+                continue;
+            }
+
+            // Try to pass the templated content resource
+            TemplatedResource templatedResource = childPageContentResource.adaptTo(TemplatedResource.class);
+
+            if (templatedResource != null) {
+                childPageContentResource = templatedResource;
+            }
+
+            itemWrappers.put(childPage.getPath(), modelFactory.getModelFromWrappedRequest(PageImpl.getHierarchyServletRequest(slingRequestWrapper, childPage), childPageContentResource, Page.class));
+        }
+
+        return itemWrappers;
+    }
+
+    /**
+     * Returns a model URL for the given page URL
+     *
+     * @param slingRequest the current servlet request
+     * @param page page for which to get the model URL
+     * @return {@link String} model URL
+     */
+    private String getModelUrl(@Nonnull SlingHttpServletRequest slingRequest, @Nonnull com.day.cq.wcm.api.Page page) {
+        return PageImpl.getModelUrl(Utils.getURL(slingRequest, page));
+    }
+
+    /**
+     * @return Returns the root (app) page the current page is part of
+     */
+    private com.day.cq.wcm.api.Page getRootPage() {
```

Comment:
This method seems expensive and it's potentially called multiple times, please cache the results after the first call!